File: legacy/python/src/limitwatch/export.py
"""Export functionality for historical quota data."""

import csv
import io
import logging
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, Optional

from .history import HistoryManager

logger = logging.getLogger(__name__)


class Exporter:
    """Exports historical quota data to various formats."""

    def __init__(self, history_mgr: Optional[HistoryManager] = None):
        self.history_mgr = history_mgr or HistoryManager()

    def export_csv(
        self,
        output_path: Optional[Path] = None,
        preset: Optional[str] = None,
        since: Optional[str] = None,
        until: Optional[str] = None,
        account_email: Optional[str] = None,
        provider_type: Optional[str] = None,
        quota_name: Optional[str] = None,
    ) -> str:
        """Export history data to CSV format.

        Args:
            output_path: Path to write CSV file (if None, returns string)
            preset: Time preset (24h, 7d, 30d, 90d)
            since: Start time string
            until: End time string
            account_email: Filter by account
            provider_type: Filter by provider
            quota_name: Filter by quota

        Returns:
            CSV string if output_path is None, otherwise empty string
        """
        history_data = self.history_mgr.get_history(
            preset=preset,
            since=since,
            until=until,
            account_email=account_email,
            provider_type=provider_type,
            quota_name=quota_name,
        )

        if not history_data:
            logger.warning("No data to export")
            return ""

        output = io.StringIO()
        writer = csv.writer(output)

        # Write header
        writer.writerow(
            [
                "timestamp",
                "account_email",
                "provider_type",
                "quota_name",
                "display_name",
                "remaining_pct",
                "used",
                "limit",
                "reset_time",
            ]
        )

        # Write data
        for record in history_data:
            writer.writerow(
                [
                    record.get("timestamp", ""),
                    record.get("account_email", ""),
                    record.get("provider_type", ""),
                    record.get("quota_name", ""),
                    record.get("display_name", ""),
                    record.get("remaining_pct", ""),
                    record.get("used", ""),
                    record.get("limit_val", ""),
                    record.get("reset_time", ""),
                ]
            )

        csv_content = output.getvalue()
        output.close()

        if output_path:
            output_path = Path(output_path)
            output_path.parent.mkdir(parents=True, exist_ok=True)
            with open(output_path, "w", newline="") as f:
                f.write(csv_content)
            logger.info(f"Exported {len(history_data)} records to {output_path}")
            return ""

        return csv_content

    def export_markdown(
        self,
        output_path: Optional[Path] = None,
        preset: Optional[str] = None,
        since: Optional[str] = None,
        until: Optional[str] = None,
        account_email: Optional[str] = None,
        provider_type: Optional[str] = None,
        quota_name: Optional[str] = None,
    ) -> str:
        """Export history data to Markdown table format.

        Args:
            output_path: Path to write Markdown file (if None, returns string)
            preset: Time preset (24h, 7d, 30d, 90d)
            since: Start time string
            until: End time string
            account_email: Filter by account
            provider_type: Filter by provider
            quota_name: Filter by quota

        Returns:
            Markdown string if output_path is None, otherwise empty string
        """
        history_data = self.history_mgr.get_history(
            preset=preset,
            since=since,
            until=until,
            account_email=account_email,
            provider_type=provider_type,
            quota_name=quota_name,
        )

        if not history_data:
            logger.warning("No data to export")
            return ""

        lines = []
        lines.append("# Quota History Export")
        lines.append("")
        lines.append(f"Generated: {datetime.now().isoformat()}")
        lines.append("")

        # Add filters info
        if preset or since or account_email or provider_type or quota_name:
            lines.append("## Filters")
            if preset:
                lines.append(f"- Time Range: {preset}")
            if since:
                lines.append(f"- Since: {since}")
            if until:
                lines.append(f"- Until: {until}")
            if account_email:
                lines.append(f"- Account: {account_email}")
            if provider_type:
                lines.append(f"- Provider: {provider_type}")
            if quota_name:
                lines.append(f"- Quota: {quota_name}")
            lines.append("")

        lines.append("## Data")
        lines.append("")
        lines.append(
            "| Timestamp | Account | Provider | Quota | Remaining % | Used | Limit |"
        )
        lines.append(
            "|-----------|---------|----------|-------|-------------|------|-------|"
        )

        for record in history_data:
            timestamp = record.get("timestamp", "")[:16]  # Truncate to remove seconds
            account = record.get("account_email", "").split("@")[0]
            provider = record.get("provider_type", "")
            quota = record.get("display_name") or record.get("quota_name", "")
            remaining = (
                f"{record.get('remaining_pct', 0):.1f}%"
                if record.get("remaining_pct") is not None
                else "N/A"
            )
            used = (
                f"{record.get('used', 0):.0f}"
                if record.get("used") is not None
                else "N/A"
            )
            limit = (
                f"{record.get('limit_val', 0):.0f}"
                if record.get("limit_val") is not None
                else "N/A"
            )

            lines.append(
                f"| {timestamp} | {account} | {provider} | {quota} | {remaining} | {used} | {limit} |"
            )

        lines.append("")
        lines.append(f"*Total records: {len(history_data)}*")
        lines.append("")

        markdown_content = "\n".join(lines)

        if output_path:
            output_path = Path(output_path)
            output_path.parent.mkdir(parents=True, exist_ok=True)
            with open(output_path, "w") as f:
                f.write(markdown_content)
            logger.info(f"Exported {len(history_data)} records to {output_path}")
            return ""

        return markdown_content

    def get_export_info(
        self,
        preset: Optional[str] = None,
        since: Optional[str] = None,
        until: Optional[str] = None,
        account_email: Optional[str] = None,
        provider_type: Optional[str] = None,
        quota_name: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get information about what would be exported.

        Returns:
            Dict with record count and date range
        """
        history_data = self.history_mgr.get_history(
            preset=preset,
            since=since,
            until=until,
            account_email=account_email,
            provider_type=provider_type,
            quota_name=quota_name,
        )

        if not history_data:
            return {
                "record_count": 0,
                "date_range": None,
            }

        timestamps = [r["timestamp"] for r in history_data if r.get("timestamp")]
        if timestamps:
            timestamps.sort()
            return {
                "record_count": len(history_data),
                "date_range": {
                    "start": timestamps[0],
                    "end": timestamps[-1],
                },
            }

        return {
            "record_count": len(history_data),
            "date_range": None,
        }

File: limitwatch-rs/src/export.rs
use crate::history::HistoryManager;
use anyhow::Result;
use chrono::Local;
use std::fmt::Write as _;
use std::path::Path;

#[derive(Default)]
pub struct ExportFilter<'a> {
    pub preset: Option<&'a str>,
    pub since: Option<&'a str>,
    pub until: Option<&'a str>,
    pub account_email: Option<&'a str>,
    pub provider_type: Option<&'a str>,
    pub quota_name: Option<&'a str>,
}
pub struct ExportInfo {
    pub record_count: usize,
    pub date_range: Option<(String, String)>,
}
pub struct Exporter<'a> {
    pub history: &'a HistoryManager,
}
impl Exporter<'_> {
    fn data(&self, f: &ExportFilter<'_>) -> Result<Vec<crate::storage::Snapshot>> {
        self.history.get_history(
            f.preset,
            f.since,
            f.until,
            f.account_email,
            f.provider_type,
            f.quota_name,
        )
    }
    fn finish(content: String, path: Option<&Path>) -> Result<String> {
        if let Some(p) = path {
            if let Some(parent) = p.parent() {
                std::fs::create_dir_all(parent)?
            }
            std::fs::write(p, content)?;
            Ok(String::new())
        } else {
            Ok(content)
        }
    }
    pub fn export_csv(&self, path: Option<&Path>, f: &ExportFilter<'_>) -> Result<String> {
        let data = self.data(f)?;
        if data.is_empty() {
            return Ok(String::new());
        }
        let mut out="timestamp,account_email,provider_type,quota_name,display_name,remaining_pct,used,limit,reset_time\r\n".to_owned();
        for x in data {
            let fields = [
                x.timestamp,
                x.account_email,
                x.provider_type,
                x.quota_name,
                x.display_name.unwrap_or_default(),
                x.remaining_pct.map(|v| v.to_string()).unwrap_or_default(),
                x.used.map(|v| v.to_string()).unwrap_or_default(),
                x.limit_val.map(|v| v.to_string()).unwrap_or_default(),
                x.reset_time.unwrap_or_default(),
            ];
            out.push_str(
                &fields
                    .iter()
                    .map(|v| csv_field(v))
                    .collect::<Vec<_>>()
                    .join(","),
            );
            out.push_str("\r\n")
        }
        Self::finish(out, path)
    }
    pub fn export_markdown(&self, path: Option<&Path>, f: &ExportFilter<'_>) -> Result<String> {
        let data = self.data(f)?;
        if data.is_empty() {
            return Ok(String::new());
        }
        let mut o = format!(
            "# Quota History Export\n\nGenerated: {}\n\n",
            Local::now().to_rfc3339()
        );
        if [
            f.preset,
            f.since,
            f.account_email,
            f.provider_type,
            f.quota_name,
        ]
        .iter()
        .any(|x| x.is_some())
        {
            o.push_str("## Filters\n");
            for (label, v) in [
                ("Time Range", f.preset),
                ("Since", f.since),
                ("Until", f.until),
                ("Account", f.account_email),
                ("Provider", f.provider_type),
                ("Quota", f.quota_name),
            ] {
                if let Some(v) = v {
                    writeln!(o, "- {label}: {v}")?
                }
            }
            o.push('\n')
        }
        o.push_str("## Data\n\n| Timestamp | Account | Provider | Quota | Remaining % | Used | Limit |\n|-----------|---------|----------|-------|-------------|------|-------|\n");
        let n = data.len();
        for x in data {
            let ts = &x.timestamp[..x.timestamp.len().min(16)];
            let account = x.account_email.split('@').next().unwrap_or("");
            let quota = x
                .display_name
                .as_deref()
                .filter(|x| !x.is_empty())
                .unwrap_or(&x.quota_name);
            let r = x
                .remaining_pct
                .map(|v| format!("{v:.1}%"))
                .unwrap_or_else(|| "N/A".into());
            let u = x
                .used
                .map(|v| format!("{v:.0}"))
                .unwrap_or_else(|| "N/A".into());
            let l = x
                .limit_val
                .map(|v| format!("{v:.0}"))
                .unwrap_or_else(|| "N/A".into());
            writeln!(
                o,
                "| {ts} | {account} | {} | {quota} | {r} | {u} | {l} |",
                x.provider_type
            )?
        }
        write!(o, "\n*Total records: {n}*\n")?;
        Self::finish(o, path)
    }
    pub fn get_export_info(&self, f: &ExportFilter<'_>) -> Result<ExportInfo> {
        let d = self.data(f)?;
        let range = if d.is_empty() {
            None
        } else {
            let mut t = d.iter().map(|x| x.timestamp.clone()).collect::<Vec<_>>();
            t.sort();
            Some((t[0].clone(), t[t.len() - 1].clone()))
        };
        Ok(ExportInfo {
            record_count: d.len(),
            date_range: range,
        })
    }
}
fn csv_field(v: &str) -> String {
    if v.chars().any(|c| matches!(c, ',' | '"' | '\n' | '\r')) {
        format!("\"{}\"", v.replace('"', "\"\""))
    } else {
        v.to_owned()
    }
}
